packages/client/hmi-client/src/services/model-templating.ts

```diff
@@ -196,95 +260,143 @@ export function addCard(
 				addTemplateArguments.subject_initial_value = initials[0].expression;
 			}
 		} else {
-			const { observables } = modelTemplate.semantics.ode;
+			const { observables } = modelTemplateToAdd.semantics.ode;
 			addTemplateArguments.new_id = observables[0].id;
-			addTemplateArguments.new_name = modelTemplate.header.name;
+			addTemplateArguments.new_name = modelTemplateToAdd.header.name;
 			addTemplateArguments.new_expression = observables[0].expression_mathml;
 		}
 
 		kernelManager
-			.sendMessage(`add_${snakeCase(name)}_template_request`, addTemplateArguments)
-			.register(`add_${snakeCase(name)}_template_response`, (d) => {
+			.sendMessage(`add_${snakeCase(templateType)}_template_request`, addTemplateArguments)
+			.register(`add_${snakeCase(templateType)}_template_response`, (d) => {
 				outputCode(d);
 			});
+
+		addTemplateInView(modelTemplates, modelTemplateToAdd);
 	}
-	// FIXME: There is some lag when placing a card if these are put in the register callback which feels off
-	// Perhaps there can be some sort of transition state to show the card is being placed?
-	modelTemplate.metadata.templateCard.id = uuidv4();
-	modelTemplates.models.push(modelTemplate);
 }
 
-export function updateDecomposedCardName(
-	modelTemplates: ModelTemplates,
+/**
+ * Remove template card
+ */
+export function removeTemplateInView(modelTemplates: ModelTemplates, id: string) {
+	const index = findCardIndexById(modelTemplates, id);
+	// Remove edges connected to the card
+	modelTemplates.junctions.forEach((junction) => {
+		junction.edges = junction.edges.filter((edge) => edge.target.cardId !== id);
+	});
+	junctionCleanUp(modelTemplates);
+	modelTemplates.models.splice(index, 1); // Remove card
+}
+
+export function removeTemplateInKernel(
 	kernelManager: KernelSessionManager,
-	outputCode: Function,
-	name: string,
-	id: string
+	modelTemplates: ModelTemplates,
+	id: string,
+	outputCode: Function
 ) {
 	const index = findCardIndexById(modelTemplates, id);
-	const model = modelTemplates.models[index];
-
 	kernelManager
-		.sendMessage('replace_template_name_request', {
-			old_name: model.header.name,
-			new_name: name
+		.sendMessage('remove_template_request', {
+			template_name: modelTemplates.models[index].metadata.templateCard.name
 		})
-		.register('replace_template_name_response', (d) => {
+		.register('remove_template_response', (d) => {
+			removeTemplateInView(modelTemplates, id);
 			outputCode(d);
-
-			model.header.name = name;
-			model.metadata.templateCard.name = name;
-			if (model.model.transitions[0] && model.semantics.ode.rates[0]) {
-				model.model.transitions[0].id = name;
-				model.model.transitions[0].properties.name = name;
-				model.semantics.ode.rates[0].target = name;
-			}
 		});
 }
 
-export function removeCard(
-	modelTemplates: ModelTemplates,
+/**
+ * Update decomposed template name
+ */
+export function updateDecomposedTemplateNameInView(model: any, newName: string) {
```

Comment:
You have a fair amount of `model: any` which I think can and should probably be typed:
- model can correspond to `types/Types#Model`, while
- model.model can, for the time, correspond to `types/Types#PetrinetModel`

Reply:
I was using any since `ModelMetadata` doesn't hold a `templateCard` property, I'll see about adding it in.

Reply:
Added the types, which made me add quite a bit of checks for properties